api/core/v1alpha1/clustercatalog_types.go

```diff
@@ -164,6 +164,9 @@ type CatalogSource struct {
 }
 
 // ResolvedCatalogSource is a discriminated union of resolution information for a Catalog.
+// ResolvedCatalogSource contains the information about a sourced Catalog
+// +union
+// +kubebuilder:validation:XValidation:rule="self.type != 'Image' || has(self.image)",message="sourceType 'Image' requires image field"
```

Comment:
I think we may want something more like:
```suggestion
// +kubebuilder:validation:XValidation:rule="self.type == 'Image' && has(self.image)",message="sourceType 'Image' requires image field"
```
This ensures that we return "True" for the validation _only if_ the type is `Image` _and_ the `image` field is set.

IIUC, the current validation of `self.type !=  'Image' || has(self.image)` returns "True" in scenarios that would be invalid:
- `type` is set to `Foo` (not currently possible based on enums, but for sake of ensuring we have a solid CEL validation assume we can)  and the `image` field is set 
- `type` is set to `Image` and the `image` field is unset

Reply:
I ran into some issues when using `self.type == 'Image' && has(self.image)`, mostly when I tried validating with an unsupported image type and when I tried adding another supported `sourceType` to test extensibility. 

the `+kubebuilder:validation:XValidation` tag requires all its validation checks to be true, so we were failing on the `self.type=='Image'` part of the check even when the source type was something else. I added the `self.type != 'Image'` so we wouldn't be doing that check on non-image sources.

I've added another validation to ensure we don't have the `image` field on non-image sources to cover the case you highlighted, thanks for pointing that out! We'll still fail on non-supported sources because of the enum check, but this won't throw the `type 'Image' requires image field` error when the source type isn't 'Image'

Reply:
>I ran into some issues when using self.type == 'Image' && has(self.image), mostly when I tried validating with an unsupported image type and when I tried adding another supported sourceType to test extensibility

I think this is something that I would expect based on our allowed enum values.

>the +kubebuilder:validation:XValidation tag requires all its validation checks to be true, so we were failing on the self.type=='Image' part of the check even when the source type was something else. I added the self.type != 'Image' so we wouldn't be doing that check on non-image sources.

I'm a bit confused here. Are you saying that it the XValidation tag doesn't evaluate the expression holistically but instead each individual expression within a chain? This sounds off to me - do you have any documentation or references you've found that state that is the case?

If I'm not mistaken, what seems more likely here is that the validations are failing on the fact that we only allow the `Image` value as part of our enum checks so we don't even hit the CEL validations.

Holistically I would expect the following test cases to fail, but not hit our CEL validations until more than the `Image` value is allowed:
- Attempting to set type to something other than `Image`
- Attempting to set type to something other than `Image` AND setting image field

For ^ I would avoid checking for specific error messages for now and just ensuring that we encounter an error.

One that I would expect us to fail from the CEL validation:
- Attempting to set type to `Image` but nil image field

One that I would expect to succeed:
- Attempting to set type to `Image` and set image field

I _do_ think we should be able to satisfy all these cases with a singular CEL expression with the current state of our API. Future iterations to the API that add new source types will likely need to update the CEL validations so I would avoid trying to accommodate future use cases of different source type values in the CEL introduced as part of this PR

Reply:
> I think this is something that I would expect based on our allowed enum values.
I'd added another allowed value to our enums and regenerated manifests to test this case
`	// +kubebuilder:validation:Enum:="Image";"someothertype"`

> Are you saying that it the XValidation tag doesn't evaluate the expression holistically but instead each individual expression within a chain?

It seems that way from https://book.kubebuilder.io/reference/markers/crd-validation
```
// +kubebuilder:validation:XValidation
    marks a field as requiring a value for which a given expression evaluates to true.
    This marker may be repeated to specify multiple expressions, all of which must evaluate to true.
```
I verified this with the test I mentioned earlier. A catalog with the type "someothertype" would fail with or without the `image` field with 
`The ClusterCatalog "testcatalogothertype" is invalid: spec.source: Invalid value: "object": source type 'Image' requires image field`

Reply:
When we add another sourceType in the future, we will have to modify this expression. e.g.
```
(self.type == 'Image' && has(self.image)) || (self.type == 'Foo' && has(self.foo))
```

Reply:
>// +kubebuilder:validation:Enum:="Image";"someothertype"

@ankitathomas I don't think this is a valid validation tag. I think that it is supposed to be:
```
// +kubebuilder:validation:Enum:="Image,SomeOtherType"
```

Reply:
> `// +kubebuilder:validation:Enum:="Image,SomeOtherType"`

That seems incorrect, the manifest generated that way considers the valid value to be "Image,SomeOtherType"

`The ClusterCatalog "operatorhubio2" is invalid: spec.source.type: Unsupported value: "SomeOtherType": supported values: "Image,someothertype"`

I also tried with `// +kubebuilder:validation:Enum:="Image","someothertype"` ,  that also doesn't work.

`// +kubebuilder:validation:Enum:="Image";"someothertype"` works correctly though, if the CEL validation is removed, the catalog manifest with the "someothertype" applies without any issues

Reply:
> (self.type == 'Image' && has(self.image)) || (self.type == 'Foo' && has(self.foo))

I think we won't be able to tailor our validation errors if we make the CEL validation one long validation for all types. We'd also be throwing the CEL validation error when we encounter an unsupported type

```
apiVersion: olm.operatorframework.io/v1alpha1
kind: ClusterCatalog
metadata:
  name: operatorhubio2
spec:
  source:
    type: unsupported

```
```
The ClusterCatalog "operatorhubio2" is invalid: 
* spec.source.type: Unsupported value: "unsupported": supported values: "Image", "Foo"
* spec: Invalid value: "object": no such key: image evaluating rule: cannot specify PollInterval while using digest-based image
* spec.source: Invalid value: "object": source type 'Image' requires image field
```
Not sure why it's throwing an error with the PollInterval though.

This was the expression set preserves a little more info about validation errors:
```
// +kubebuilder:validation:XValidation:rule="self.type != 'Image' || has(self.image)",message="source type 'Image' requires image field"
// +kubebuilder:validation:XValidation:rule="self.type == 'Image' || !has(self.image)",message="image field must only be set for source type 'Image'"
// +kubebuilder:validation:XValidation:rule="self.type != 'Foo' || has(self.foo)",message="source type 'Foo' requires foo field"
// +kubebuilder:validation:XValidation:rule="self.type == 'Foo' || !has(self.foo)",message="foo field must only be set for source type 'Foo'"
```

an unsuported type gives the following error set:
```
The ClusterCatalog "operatorhubio2" is invalid: 
* spec.source.type: Unsupported value: "unsupported": supported values: "Image", "Foo"
* spec: Invalid value: "object": no such key: image evaluating rule: cannot specify PollInterval while using digest-based image
```

A catalog with an invalid field gives something like:
```
kind: ClusterCatalog
metadata:
  name: operatorhubio2
spec:
  source:
    type: Foo
    image:
      ref: ""
```
```
The ClusterCatalog "operatorhubio2" is invalid: 
* spec.source: Invalid value: "object": image field must only be set for source type 'Image'
* spec.source: Invalid value: "object": source type 'Foo' requires foo field
```

We can combine the validations for each type so we don't end up too verbose
```
// +kubebuilder:validation:XValidation:rule="(self.type != 'Image' || has(self.image)) && (self.type == 'Image' || !has(self.image))",message="source type 'Image' requires image field"
// +kubebuilder:validation:XValidation:rule="(self.type != 'Foo' && has(self.foo)) || (self.type == 'Foo' && has(self.foo))",message="source type 'Foo' requires foo field"
```

Reply:
Ah I didn't realize we could have multiple validation rules like that, but I guess it makes sense. That's nicer because we can tailor the message for each one.

If we want to consolidate to a single rule, we can use `messageExpression` instead of `message` in order to use CEL to build a message. But I'm not sure that would give us the flexibility we could need to craft a helpful validation message.

Reply:
I think having multiple validation rules like that is fine, but I'd be curious how that translates to the CEL cost budget.

I'm still a bit hung up on the OR comparisons we are doing here to verify the cases. Maybe a sync would be helpful here?

Reply:
follow up: https://github.com/operator-framework/catalogd/pull/384#discussion_r1783254070